### PR TITLE
support for services

### PR DESCRIPTION
**Pull request by lpuskas**
!!!!! Braking change - don't merge yet! - early review

- added new url/REST scheme
- new endpoints for supporting services
- skeletons for supporting services in provider specific code (provider specific implementations to be added)

TBD:
- swagger annotations / regenerate client
- regenerate api spec

### CODE REVIEW

File: internal/app/productinfo/api/handlers.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import (
+	"net/http"
+
+	"fmt"
+	"github.com/gin-gonic/gin"
+	log "github.com/sirupsen/logrus"
+)
+
+// getServices retrieves services supported by the gibven provider in the given region
```
Comment (sagikazarmark) [inline internal/app/productinfo/api/handlers.go]: Typo: `gibven`

File: internal/app/productinfo/api/handlers.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+func (r *RouteHandler) getServices(c *gin.Context) {
+
+	pathParams := getPathParams(c)
+	log.Debug("Path params: [%s]", pathParams)
+
+	var err error
```
Comment (sagikazarmark) [inline internal/app/productinfo/api/handlers.go]: This error will never have a value and will always be nil. The redeclared `err` variable on the following line will only be valid within the if scope.

Also, in Go the happy path usually ends at the end of the function and error cases are handled earlier. So in this case errors should be checked for and handled first, like this:

```go
infoer, err := r.prod.GetInfoer(pathParams.Provider)
if err != nil {
    c.JSON(http.StatusInternalServerError, fmt.Sprintf("error while retrieving services: %s", err.Error()))

    return
}

services, err := infoer.GetServices(pathParams.Region)
if err != nil {
    c.JSON(http.StatusInternalServerError, fmt.Sprintf("error while retrieving services: %s", err.Error()))
    
    return
}

c.JSON(http.StatusOK, services)
```

Code duplication is less of an issue here.

File: internal/app/productinfo/api/handlers.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	c.JSON(http.StatusInternalServerError, fmt.Sprintf("error while retrieving services: %s", err.Error()))
+
+}
+
+func (r *RouteHandler) getService(c *gin.Context) {
+
+	pathParams := getPathParams(c)
+	log.Debug("Path params: [%s]", pathParams)
+
+	c.JSON(http.StatusInternalServerError, "not yet implemented")
```
Comment (sagikazarmark) [inline internal/app/productinfo/api/handlers.go]: Just a semantical thing, but there is an http status code for this: 501 Not Implemented

File: internal/app/productinfo/api/handlers.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+}
+
+func (r *RouteHandler) getServiceImages(c *gin.Context) {
+
+	pathParams := getPathParams(c)
+	log.Debug("Path params: [%s]", pathParams)
+
+	var err error
```
Comment (sagikazarmark) [inline internal/app/productinfo/api/handlers.go]: Same as `getServices`

File: internal/app/productinfo/api/handlers.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	c.JSON(http.StatusInternalServerError, fmt.Sprintf("error while retrieving service images: %s", err.Error()))
+
+}
+
+func (r *RouteHandler) getServiceProducts(c *gin.Context) {
+
+	pathParams := getPathParams(c)
+	log.Debug("Path params: [%s]", pathParams)
+
+	var err error
```
Comment (sagikazarmark) [inline internal/app/productinfo/api/handlers.go]: Same as `getServices`

File: internal/app/productinfo/api/handlers.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+	c.JSON(http.StatusInternalServerError, fmt.Sprintf("error while retrieving service products: %s", err.Error()))
+
+}
+
+func (r *RouteHandler) getServiceAttributes(c *gin.Context) {
+	pathParams := getPathParams(c)
+	log.Debug("Path params: [%s]", pathParams)
+
+	var err error
```
Comment (sagikazarmark) [inline internal/app/productinfo/api/handlers.go]: Same as `getServices`

**Comment by lpuskas** [comment]
closing this as it's pretty outdated, will open a new one with more recent changeset

**Pull request closed without merging**